File: frontend/web-app/src/app/core/services/project.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';
import {
  Project,
  ProjectTask,
  Milestone,
  TimeEntry,
  ProjectDocument,
  ProjectStats,
  GanttTask,
  BurndownData,
  ProjectStatus,
  TaskStatus
} from '../models/project.model';

@Injectable({ providedIn: 'root' })
export class ProjectService {
  private readonly API_URL = `${environment.apiUrl}/projects`;

  constructor(private http: HttpClient) {}

  // Projects
  getProjects(status?: ProjectStatus, managerId?: string): Observable<Project[]> {
    let params = new HttpParams();
    if (status) params = params.set('status', status);
    if (managerId) params = params.set('managerId', managerId);
    return this.http.get<Project[]>(`${this.API_URL}`, { params });
  }

  getProject(id: string): Observable<Project> {
    return this.http.get<Project>(`${this.API_URL}/${id}`);
  }

  createProject(project: Partial<Project>): Observable<Project> {
    return this.http.post<Project>(`${this.API_URL}`, project);
  }

  updateProject(id: string, project: Partial<Project>): Observable<Project> {
    return this.http.put<Project>(`${this.API_URL}/${id}`, project);
  }

  deleteProject(id: string): Observable<void> {
    return this.http.delete<void>(`${this.API_URL}/${id}`);
  }

  // Tasks
  getTasks(projectId: string): Observable<ProjectTask[]> {
    return this.http.get<ProjectTask[]>(`${this.API_URL}/${projectId}/tasks`);
  }

  createTask(projectId: string, task: Partial<ProjectTask>): Observable<ProjectTask> {
    return this.http.post<ProjectTask>(`${this.API_URL}/${projectId}/tasks`, task);
  }

  updateTask(projectId: string, taskId: string, task: Partial<ProjectTask>): Observable<ProjectTask> {
    return this.http.put<ProjectTask>(`${this.API_URL}/${projectId}/tasks/${taskId}`, task);
  }

  deleteTask(projectId: string, taskId: string): Observable<void> {
    return this.http.delete<void>(`${this.API_URL}/${projectId}/tasks/${taskId}`);
  }

  // Milestones
  getMilestones(projectId: string): Observable<Milestone[]> {
    return this.http.get<Milestone[]>(`${this.API_URL}/${projectId}/milestones`);
  }

  createMilestone(projectId: string, milestone: Partial<Milestone>): Observable<Milestone> {
    return this.http.post<Milestone>(`${this.API_URL}/${projectId}/milestones`, milestone);
  }

  // Time Entries
  logTime(projectId: string, entry: Partial<TimeEntry>): Observable<TimeEntry> {
    return this.http.post<TimeEntry>(`${this.API_URL}/${projectId}/time-entries`, entry);
  }

  getTimeReport(projectId: string): Observable<any> {
    return this.http.get<any>(`${this.API_URL}/${projectId}/time-report`);
  }

  // Gantt
  getGanttData(projectId: string): Observable<GanttTask[]> {
    return this.http.get<GanttTask[]>(`${this.API_URL}/${projectId}/gantt`);
  }

  // Burndown
  getBurndownData(projectId: string): Observable<BurndownData[]> {
    return this.http.get<BurndownData[]>(`${this.API_URL}/${projectId}/burndown`);
  }

  // Documents
  getDocuments(projectId: string): Observable<ProjectDocument[]> {
    return this.http.get<ProjectDocument[]>(`${this.API_URL}/${projectId}/documents`);
  }

  // Budget
  getBudget(projectId: string): Observable<{ allocated: number; spent: number; remaining: number }> {
    return this.http.get<any>(`${this.API_URL}/${projectId}/budget`);
  }

  // Stats
  getStats(): Observable<ProjectStats> {
    return this.http.get<ProjectStats>(`${this.API_URL}/stats`);
  }
}

File: backend/project-service/src/main/java/ro/pdi/project/service/ProjectService.java
package ro.pdi.project.service;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import ro.pdi.project.model.*;
import ro.pdi.project.repository.*;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;

@Service
@RequiredArgsConstructor
@Slf4j
public class ProjectService {

    private final ProjectRepository projectRepository;
    private final ProjectTaskRepository taskRepository;
    private final MilestoneRepository milestoneRepository;
    private final TimeEntryRepository timeEntryRepository;
    private final ProjectDocumentRepository documentRepository;

    // === Projects ===
    public List<Project> getProjects(String status, UUID managerId) {
        if (status != null && !status.isEmpty()) {
            return projectRepository.findByStatusAndIsActiveTrue(status);
        }
        if (managerId != null) {
            return projectRepository.findByManagerIdAndIsActiveTrue(managerId);
        }
        return projectRepository.findByIsActiveTrue();
    }

    public Project getProjectById(UUID id) {
        return projectRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Project not found: " + id));
    }

    @Transactional
    public Project createProject(Project project) {
        log.info("Creating project: {}", project.getName());
        return projectRepository.save(project);
    }

    @Transactional
    public Project updateProject(UUID id, Project updated) {
        log.info("Updating project: {}", id);
        Project existing = getProjectById(id);
        existing.setName(updated.getName());
        existing.setDescription(updated.getDescription());
        existing.setStatus(updated.getStatus());
        existing.setPriority(updated.getPriority());
        existing.setStartDate(updated.getStartDate());
        existing.setEndDate(updated.getEndDate());
        existing.setProgress(updated.getProgress());
        existing.setManagerId(updated.getManagerId());
        existing.setManagerName(updated.getManagerName());
        existing.setBudget(updated.getBudget());
        return projectRepository.save(existing);
    }

    @Transactional
    public void deleteProject(UUID id) {
        log.info("Soft-deleting project: {}", id);
        Project project = getProjectById(id);
        project.setIsActive(false);
        projectRepository.save(project);
    }

    // === Tasks ===
    public List<ProjectTask> getTasks(UUID projectId) {
        return taskRepository.findByProjectIdAndIsActiveTrue(projectId);
    }

    public ProjectTask getTaskById(UUID id) {
        return taskRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Task not found: " + id));
    }

    @Transactional
    public ProjectTask createTask(ProjectTask task) {
        log.info("Creating task: {}", task.getTitle());
        return taskRepository.save(task);
    }

    @Transactional
    public ProjectTask updateTask(UUID id, ProjectTask updated) {
        log.info("Updating task: {}", id);
        ProjectTask existing = getTaskById(id);
        existing.setTitle(updated.getTitle());
        existing.setDescription(updated.getDescription());
        existing.setStatus(updated.getStatus());
        existing.setPriority(updated.getPriority());
        existing.setAssigneeId(updated.getAssigneeId());
        existing.setAssigneeName(updated.getAssigneeName());
        existing.setStartDate(updated.getStartDate());
        existing.setDueDate(updated.getDueDate());
        existing.setEstimatedHours(updated.getEstimatedHours());
        existing.setActualHours(updated.getActualHours());
        existing.setProgress(updated.getProgress());
        return taskRepository.save(existing);
    }

    @Transactional
    public void deleteTask(UUID id) {
        log.info("Soft-deleting task: {}", id);
        ProjectTask task = getTaskById(id);
        task.setIsActive(false);
        taskRepository.save(task);
    }

    // === Milestones ===
    public List<Milestone> getMilestones(UUID projectId) {
        return milestoneRepository.findByProjectIdAndIsActiveTrue(projectId);
    }

    public Milestone getMilestoneById(UUID id) {
        return milestoneRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Milestone not found: " + id));
    }

    @Transactional
    public Milestone createMilestone(Milestone milestone) {
        log.info("Creating milestone: {}", milestone.getName());
        return milestoneRepository.save(milestone);
    }

    @Transactional
    public Milestone updateMilestone(UUID id, Milestone updated) {
        log.info("Updating milestone: {}", id);
        Milestone existing = getMilestoneById(id);
        existing.setName(updated.getName());
        existing.setDescription(updated.getDescription());
        existing.setDueDate(updated.getDueDate());
        existing.setStatus(updated.getStatus());
        return milestoneRepository.save(existing);
    }

    @Transactional
    public void deleteMilestone(UUID id) {
        log.info("Soft-deleting milestone: {}", id);
        Milestone milestone = getMilestoneById(id);
        milestone.setIsActive(false);
        milestoneRepository.save(milestone);
    }

    // === Time Entries ===
    public List<TimeEntry> getTimeEntries(UUID projectId) {
        return timeEntryRepository.findByProjectIdAndIsActiveTrue(projectId);
    }

    public List<TimeEntry> getTimeEntriesByTask(UUID taskId) {
        return timeEntryRepository.findByTaskIdAndIsActiveTrue(taskId);
    }

    public TimeEntry getTimeEntryById(UUID id) {
        return timeEntryRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Time entry not found: " + id));
    }

    @Transactional
    public TimeEntry createTimeEntry(TimeEntry timeEntry) {
        log.info("Creating time entry for project: {}", timeEntry.getProjectId());
        return timeEntryRepository.save(timeEntry);
    }

    @Transactional
    public void deleteTimeEntry(UUID id) {
        log.info("Soft-deleting time entry: {}", id);
        TimeEntry timeEntry = getTimeEntryById(id);
        timeEntry.setIsActive(false);
        timeEntryRepository.save(timeEntry);
    }

    // === Documents ===
    public List<ProjectDocument> getDocuments(UUID projectId) {
        return documentRepository.findByProjectIdAndIsActiveTrue(projectId);
    }

    public ProjectDocument getDocumentById(UUID id) {
        return documentRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Document not found: " + id));
    }

    @Transactional
    public ProjectDocument createDocument(ProjectDocument document) {
        log.info("Creating document: {}", document.getTitle());
        return documentRepository.save(document);
    }

    @Transactional
    public ProjectDocument updateDocument(UUID id, ProjectDocument updated) {
        log.info("Updating document: {}", id);
        ProjectDocument existing = getDocumentById(id);
        existing.setTitle(updated.getTitle());
        existing.setFileName(updated.getFileName());
        existing.setFileSize(updated.getFileSize());
        existing.setStatus(updated.getStatus());
        existing.setUploadedAt(updated.getUploadedAt());
        return documentRepository.save(existing);
    }

    @Transactional
    public void deleteDocument(UUID id) {
        log.info("Soft-deleting document: {}", id);
        ProjectDocument document = getDocumentById(id);
        document.setIsActive(false);
        documentRepository.save(document);
    }

    // === Stats ===
    public Map<String, Object> getStats() {
        Map<String, Object> stats = new HashMap<>();
        stats.put("totalProjects", projectRepository.countByIsActiveTrue());
        stats.put("totalTasks", taskRepository.count());
        stats.put("totalMilestones", milestoneRepository.count());
        stats.put("totalTimeEntries", timeEntryRepository.count());
        stats.put("totalDocuments", documentRepository.count());
        return stats;
    }
}
